### PR TITLE
Localauth API and Web UI changes

### PR DESCRIPTION
**Pull request by thomasmmitchell**
In preparation for CLI support

### CODE REVIEW

File: core/auth.go [hunk 1]
```diff
@@ -56,25 +56,34 @@ type authResponse struct {
...
-			log.Debugf("no session cookie ('%s') found in request; treating as unauthenticated", SessionCookieName)
-		} else {
-			log.Debugf("failed to extract session cookie ('%s') from request: %s", SessionCookieName, err)
+//Gets the session ID from the request. Returns "" if not given.
+func getSessionID(req *http.Request) string {
+	sessionID := req.Header.Get("X-Shield-Session")
+
+	if sessionID == "" { //If not in header, check for cookie
+		cookie, err := req.Cookie(SessionCookieName)
+		if err == nil { //ErrNoCookie is the only error returned from Cookie()
```
Comment (jhunt) [inline core/auth.go]: I would believe that now, but in the future?

File: core/v2.go [hunk 2]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+			Username string
+			Password string
+		}{}
+
+		if !r.Payload(&auth) { //Payload reports its own errors
+			return
+		}
+
+		if auth.Username == "" {
+			r.Fail(route.Errorf(http.StatusForbidden, nil, "no username given"))
```
Comment (jhunt) [inline core/v2.go]: Use `403` here

File: core/v2.go [hunk 3]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+			return
+		}
+
+		if auth.Password == "" {
+			r.Fail(route.Errorf(http.StatusForbidden, nil, "no password given"))
```
Comment (jhunt) [inline core/v2.go]: Use `403` here

File: core/v2.go [hunk 4]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+		}
+
+		user, err := core.DB.GetUser(auth.Username, "local")
+		if err != nil {
+			r.Fail(route.Oops(err, "An unknown error occurred when authenticating local user '%s'", auth.Username))
+			return
+		}
+
+		if user == nil || !user.Authenticate(auth.Password) {
+			r.Fail(route.Errorf(http.StatusForbidden, nil, "Incorrect username or password"))
```
Comment (jhunt) [inline core/v2.go]: also here

File: core/v2.go [hunk 5]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+		r.SetCookie(SessionCookie(session.UUID.String(), true))
+		r.SetHeader("X-Shield-Session", session.UUID.String())
+
+		r.Success("Logged in as user %s", auth.Username)
+	})
+
+	r.Dispatch("GET /v2/auth/logout", func(r *route.Request) {
+		sessionID := getSessionID(r.Req)
+		if sessionID == "" {
+			//I guess we're okay with not getting a session to logout?...
```
Comment (jhunt) [inline core/v2.go]: Can we be a little more assertive here?

File: core/v2.go [hunk 6]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+			r.Success("No user to logout")
+		}
+
+		id := uuid.Parse(sessionID)
+		if id == nil {
+			r.Fail(route.Bad(nil, "Invalid session ID received"))
+		}
+		err := core.DB.ClearSession(id)
+		if err != nil {
+			r.Fail(route.Oops(err, "An unknown error occurred when logging out of the session"))
```
Comment (jhunt) [inline core/v2.go]: Drop everything after "occurred"

File: core/v2.go [hunk 7]
```diff
@@ -698,6 +699,84 @@ func (core *Core) v2API() *route.Router {
...
+	r.Dispatch("GET /v2/auth/logout", func(r *route.Request) {
+		sessionID := getSessionID(r.Req)
+		if sessionID == "" {
+			//I guess we're okay with not getting a session to logout?...
+			r.Success("No user to logout")
+		}
+
+		id := uuid.Parse(sessionID)
+		if id == nil {
+			r.Fail(route.Bad(nil, "Invalid session ID received"))
```
Comment (jhunt) [inline core/v2.go]: We should probably formulate an `error` object to pass to `Bad()`, and make the FE error message more generic, like "Unable to log out"

File: web2/htdocs/index.html [hunk 8]
```diff
@@ -1433,10 +1435,24 @@
         (function () {
           api({
             type: 'GET',
-            url:  '/v2/auth/providers',
+            url: '/v2/auth/providers',
```
Comment (jhunt) [inline web2/htdocs/index.html]: Fix the formatting regression here.

File: web2/htdocs/index.html [hunk 9]
```diff
@@ -1433,10 +1435,24 @@
...
             success: function (data) {
-              $('#viewport').html(template('login', { oauth: data }));
-            },
+              $('#viewport').html(template('login', { oauth: data }))
+                .on("submit", "form", function (event) {
+                event.preventDefault()
+                api({
+                  url: "/v2/auth/login",
+                  method: "POST",
```
Comment (jhunt) [inline web2/htdocs/index.html]: This should be `type` not `method`, and it should be listed first.

File: web2/htdocs/index.html [hunk 10]
```diff
@@ -1433,10 +1435,24 @@
...
+                  data: $(event.target).serializeObject(),
+                  success: function () {
+                    document.location.href = "/";
```
Comment (jhunt) [inline web2/htdocs/index.html]: Make this a `goto("")` instead of messing with location directly

File: web2/htdocs/index.html [hunk 11]
```diff
@@ -1445,6 +1461,25 @@
         break; /* #!/login */
         // }}}
 
+      case "#!/logout": // {{{
+        (function () {
+          api({
+            url: "/v2/auth/logout",
```
Comment (jhunt) [inline web2/htdocs/index.html]: You need a `type` here, of `GET`

File: web2/htdocs/index.html [hunk 12]
```diff
@@ -1445,6 +1461,25 @@
...
+            success: function () {
+              document.location.href = "/";
```
Comment (jhunt) [inline web2/htdocs/index.html]: Use `goto("")` instead of changing document.location directly

File: web2/htdocs/index.html [hunk 13]
```diff
@@ -1445,6 +1461,25 @@
...
+            },
+            error: function (xhr) {
+              if (xhr.status >= 500) {
+                $('#viewport').html(template('BOOM'));
+              } else {
+                document.location.href = "/";
```
Comment (jhunt) [inline web2/htdocs/index.html]: Use `goto("")`

File: core/auth.go (hunk 1 above)
Comment (thomasmmitchell) [reply core/auth.go]: "Cookie returns the named cookie provided in the request or ErrNoCookie if not found. "

File: web2/htdocs/index.html (hunk 11 above)
Comment (thomasmmitchell) [reply web2/htdocs/index.html]: Are we changing the behavior of api in the future? GET is the documented default method of .ajax

Comment (jhunt) [reply web2/htdocs/index.html]: It's a good idea to be explicit when it aids understanding.

**Pull request merged**